sm_ipt/sm_ipt.c

```diff
@@ -0,0 +1,453 @@
+/*
+ * Copyright (c) 2021 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-BSD-5-Clause-Nordic
+ */
+
+
+#include <stdint.h>
+#include <stddef.h>
+#include <stdbool.h>
+
+#include "sm_ipt.h"
+#include "sm_ipt_log.h"
+#include "sm_ipt_errno.h"
+
+
+#define FLAG_RELEASE 0x80
+
+#define WORD_SIZE sizeof(uint32_t)
+
+#if defined(CONFIG_SM_IPT_NUM_BLOCKS_32)
+typedef uint32_t mask_t;
+typedef int32_t smask_t;
+#define NUM_BLOCKS 32
+#define mask_clz sm_ipt_os_clz32
+#elif defined(CONFIG_SM_IPT_NUM_BLOCKS_64)
+typedef uint64_t mask_t;
+typedef int64_t smask_t;
+#define NUM_BLOCKS 64
+#define mask_clz sm_ipt_os_clz64
+#else
+#error Number of shared dynamic memory blocks is not configured.
+#endif
+
+#define ALLOCABLE_MULTIPLY (WORD_SIZE * NUM_BLOCKS)
+#define QUEUE_INDEX_MASK (NUM_BLOCKS - 1)
+
+/* Worst case scenario: NUM_BLOCKS count of messages waiting to be sent and NUM_BLOCKS count of
+ * buffers waiting to be freed + 1 item used as indicator of full queue
```

Comment:
It is not `indicator`. Maybe: `+ 1 item that must be always free in circular buffer`